File: LeetCode-CN/3.无重复字符的最长子串.java
/*
 * @lc app=leetcode.cn id=3 lang=java
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int freq[256] = {0};
        int l = 0, r = -1; 
        int res = 0;
        while(l < s.size()){
            if(r + 1 < s.size() && freq[s[r+1]] == 0){
                r++;
                freq[s[r]]++;
            }else {
                freq[s[l]]--;
                l++;
            }
            res = max(res, r-l+1);
        }
        return res;
    }
};
// @lc code=end

File: LeetCode-C++/3.无重复字符的最长子串.cpp
/*
 * @lc app=leetcode.cn id=3 lang=cpp
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        int ans = 0;
        for(int i = 0; i < s.size(); i++) {
            unordered_map<char, int> seen;
            int j = i;
            while(j < n && seen[s[j]] == 0) {
                seen[s[j]] = 1;
                j++;
            }
            ans = max(ans, j - i);
        }
        return ans;
    }
};
// @lc code=end

